src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+            if let Some(account_key_pair) =
+                account_keys.keys.get(&CredentialIndex { index: outer_key })
+            {
+                let public_keys = account_key_pair.get_public_keys();
+
+                // Check that the length of the inner map matches.
+                if credential_sigs.sigs.len() != public_keys.len() {
+                    return Err(SignatureError::MismatchMapIndices);
+                }
+
+                // Create the zipped inner map.
+                let inner_map: Result<
+                    BTreeMap<u8, AccountSignaturesVerificationEntry>,
+                    SignatureError,
+                > = credential_sigs
+                    .sigs
+                    .iter()
+                    .zip(public_keys.iter())
+                    .map(|((&inner_key, signature), (key_index, public_key))| {
+                        // Ensure that inner_key and key_index matches.
+                        if inner_key != key_index.0 {
+                            return Err(SignatureError::MismatchMapIndices);
+                        }
+                        Ok((inner_key, AccountSignaturesVerificationEntry {
+                            signature:  signature.clone(),
+                            public_key: public_key.clone(),
+                        }))
+                    })
+                    .collect();
+
+                outer_map.insert(outer_key, CredentialSignaturesVerificationData {
+                    data: inner_map?,
+                });
+            } else {
+                return Err(SignatureError::MismatchMapIndices);
+            }
+        }
+
+        Ok(AccountSignaturesVerificationData { data: outer_map })
+    }
+}
+
+/// Calculate the message hash that is signed in Concordium wallets.
+pub fn calculate_message_hash(message: &Message<'_>, signer: AccountAddress) -> [u8; 32] {
+    let message_signed_in_wallet = match message {
+        Message::BinaryMessage(message) => message,
+        Message::TextMessage(message) => (**message).as_bytes(),
+    };
+
+    // A message signed in a Concordium wallet is prepended with the
+    // `account` address (signer) and 8 zero bytes. Accounts in a Concordium
+    // wallet can either sign a regular transaction (in that case the
+    // prepend is `account` address and the nonce of the account which is by
+    // design >= 1) or sign a message (in that case the prepend is `account`
+    // address and 8 zero bytes). Hence, the 8 zero bytes ensure that the user
+    // does not accidentally sign a transaction. The account nonce is of type
+    // u64 (8 bytes).
+    sha2::Sha256::digest(
+        [
+            &signer.as_ref() as &[u8],
+            &[0u8; 8],
+            &message_signed_in_wallet,
+        ]
+        .concat(),
+    )
+    .into()
+}
+
+/// Check that all key indices in the signatures map exist on chain but not vice
+/// versa.
+fn assert_signature_map_key_indices_on_chain<C: Curve, AttributeType: Attribute<C::Scalar>>(
```

Comment:
I would use a different word than "assert" here - `assert` in my mind implies that this function will panic if the assertion fails (like the `assert` macro), not simply return an error. Consider `verify` or `check`.